File: 875-koko-eating-bananas/875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        def bsearch(k: int) -> bool:
            hcount = 0
            for pile in piles:
                hcount += math.ceil(pile / k)
            return hcount <= h
        
        lo, hi = 1, max(piles)
        while lo < hi:
            mid = lo + (hi - lo) // 2
            
            if bsearch(mid):
                hi = mid
            else:
                lo = mid + 1
        return lo

File: 875-koko-eating-bananas/875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int maxPile = 0;
        for(int pile : piles)
            maxPile = Math.max(maxPile, pile);
        
        int lo = 1, hi = maxPile;
        while(lo < hi){
            int mid = lo + (hi - lo) / 2;
            
            int hours = 0;
            for(int pile : piles)
                hours += (pile - 1) / mid + 1;
            
            if(hours > h)
                lo = mid + 1;
            else
                hi = mid;
        }
        return lo;
    }
}